File: product-missing-default-category-after-deletion/node/fix-default-category.js
/**
 * Repair PrestaShop products left with a dangling id_category_default after a
 * category deletion.
 *
 * DeleteCategoryHandler only reassigns a product's categories when the deletion
 * would leave it with zero categories at all. It never checks whether the
 * deleted category was that product's default while the product still has
 * other valid categories, so id_category_default keeps pointing at a category
 * id that no longer exists in ps_category (PrestaShop/PrestaShop issue #30219,
 * and related issues #28016 and #9811).
 *
 * This script builds the set of valid category ids, walks every product, and
 * runs a pure decision function that picks a replacement default from the
 * product's own remaining valid categories, falling back to the shop's root
 * category. It logs every proposed change. A corrective PUT that resends the
 * full product body is only sent when DRY_RUN=false.
 *
 * Guide: https://www.allanninal.dev/prestashop/product-missing-default-category-after-deletion/
 */
import { pathToFileURL } from "node:url";

const PRESTASHOP_URL = (process.env.PRESTASHOP_URL || "https://demo.example.com").replace(/\/+$/, "");
const PRESTASHOP_WS_KEY = process.env.PRESTASHOP_WS_KEY || "WSKEYDUMMY";
const FALLBACK_ROOT_CATEGORY_ID = Number(process.env.FALLBACK_ROOT_CATEGORY_ID || 2);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/**
 * Pure decision function, no I/O.
 *
 * productId: number
 * currentDefaultId: number, the product's current id_category_default
 * associatedCategoryIds: number[], the product's full category list
 * validCategoryIds: Set<number>, every category id that still exists
 * fallbackRootId: number, used only when the product has no valid
 *   categories of its own left
 *
 * Returns an object describing what to do. action is "none" when the current
 * default is already valid, "reassign" when a safe replacement was found, or
 * "flag_manual" when no valid category exists to fall back to.
 */
export function chooseValidDefaultCategory(productId, currentDefaultId, associatedCategoryIds,
                                            validCategoryIds, fallbackRootId = 2) {
  if (validCategoryIds.has(currentDefaultId)) {
    return { id_product: productId, action: "none", new_default: currentDefaultId };
  }

  const candidates = associatedCategoryIds.filter(
    (cid) => validCategoryIds.has(cid) && cid !== currentDefaultId
  );
  const newDefault = candidates.length
    ? Math.max(...candidates)
    : (validCategoryIds.has(fallbackRootId) ? fallbackRootId : null);

  return {
    id_product: productId,
    action: newDefault ? "reassign" : "flag_manual",
    old_default: currentDefaultId,
    new_default: newDefault,
  };
}

function basicAuthHeader() {
  return "Basic " + Buffer.from(`${PRESTASHOP_WS_KEY}:`).toString("base64");
}

async function apiGet(path, params = {}) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: basicAuthHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on GET ${path}`);
  return res.json();
}

async function apiPut(path, resourceKey, body) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, {
    method: "PUT",
    headers: { Authorization: basicAuthHeader(), "Content-Type": "application/json" },
    body: JSON.stringify({ [resourceKey]: body }),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on PUT ${path}`);
  return res.json();
}

async function allCategoryIds(pageSize = 100) {
  const ids = new Set();
  let offset = 0;
  while (true) {
    const data = await apiGet("categories", { display: "full", limit: `${offset},${pageSize}` });
    const rows = data.categories || [];
    if (!rows.length) return ids;
    for (const row of rows) ids.add(Number(row.id));
    offset += pageSize;
  }
}

async function* allProducts(pageSize = 100) {
  let offset = 0;
  while (true) {
    const data = await apiGet("products", { display: "full", limit: `${offset},${pageSize}` });
    const rows = data.products || [];
    if (!rows.length) return;
    for (const row of rows) yield row;
    offset += pageSize;
  }
}

function associatedCategoryIds(product) {
  const categories = product.associations?.categories?.category || [];
  return categories.map((row) => Number(row.id));
}

async function repairProductDefaultCategory(productId, newDefaultId) {
  const data = await apiGet(`products/${productId}`);
  const product = data.product;
  product.id_category_default = newDefaultId;

  product.associations = product.associations || {};
  product.associations.categories = product.associations.categories || {};
  const rows = (product.associations.categories.category = product.associations.categories.category || []);
  if (!rows.some((row) => Number(row.id) === newDefaultId)) {
    rows.push({ id: newDefaultId });
  }

  return apiPut(`products/${productId}`, "product", product);
}

export async function run() {
  const validCategoryIds = await allCategoryIds();
  let reassigned = 0;
  let flagged = 0;
  for await (const product of allProducts()) {
    const productId = Number(product.id);
    const currentDefaultId = Number(product.id_category_default || 0);
    const decision = chooseValidDefaultCategory(
      productId, currentDefaultId, associatedCategoryIds(product),
      validCategoryIds, FALLBACK_ROOT_CATEGORY_ID,
    );
    if (decision.action === "none") continue;
    if (decision.action === "flag_manual") {
      flagged++;
      console.warn(`Product id=${productId} has no valid category to fall back to. Needs manual review.`);
      continue;
    }

    console.log(
      `Product id=${productId} old id_category_default=${decision.old_default} new id_category_default=${decision.new_default}. ${DRY_RUN ? "would reassign" : "reassigning"}`
    );
    if (!DRY_RUN) await repairProductDefaultCategory(productId, decision.new_default);
    reassigned++;
  }
  console.log(`Done. ${reassigned} product(s) ${DRY_RUN ? "to reassign" : "reassigned"}, ${flagged} flagged for manual review.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: product-missing-default-category-after-deletion/python/fix_default_category.py
"""Repair PrestaShop products left with a dangling id_category_default after a
category deletion.

DeleteCategoryHandler only reassigns a product's categories when the deletion
would leave it with zero categories at all. It never checks whether the
deleted category was that product's default while the product still has
other valid categories, so id_category_default keeps pointing at a category
id that no longer exists in ps_category (PrestaShop/PrestaShop issue #30219,
and related issues #28016 and #9811).

This script builds the set of valid category ids, walks every product, and
runs a pure decision function that picks a replacement default from the
product's own remaining valid categories, falling back to the shop's root
category. It logs every proposed change. A corrective PUT that resends the
full product body is only sent when DRY_RUN=false.

Guide: https://www.allanninal.dev/prestashop/product-missing-default-category-after-deletion/

Run on a schedule, or right after cleaning up the category tree. Safe to run
again and again.
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("fix_default_category")

PRESTASHOP_URL = os.environ["PRESTASHOP_URL"].rstrip("/")
PRESTASHOP_WS_KEY = os.environ["PRESTASHOP_WS_KEY"]
FALLBACK_ROOT_CATEGORY_ID = int(os.environ.get("FALLBACK_ROOT_CATEGORY_ID", "2"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
AUTH = (PRESTASHOP_WS_KEY, "")


def choose_valid_default_category(product_id, current_default_id, associated_category_ids,
                                   valid_category_ids, fallback_root_id=2):
    """Pure decision function, no I/O.

    product_id: int
    current_default_id: int, the product's current id_category_default
    associated_category_ids: list[int], the product's full category list
    valid_category_ids: set[int], every category id that still exists
    fallback_root_id: int, used only when the product has no valid
        categories of its own left

    Returns a dict describing what to do. action is "none" when the current
    default is already valid, "reassign" when a safe replacement was found,
    or "flag_manual" when no valid category exists to fall back to.
    """
    if current_default_id in valid_category_ids:
        return {"id_product": product_id, "action": "none", "new_default": current_default_id}

    candidates = [
        cid for cid in associated_category_ids
        if cid in valid_category_ids and cid != current_default_id
    ]
    if candidates:
        new_default = max(candidates)
    else:
        new_default = fallback_root_id if fallback_root_id in valid_category_ids else None

    return {
        "id_product": product_id,
        "action": "reassign" if new_default else "flag_manual",
        "old_default": current_default_id,
        "new_default": new_default,
    }


def api_get(path, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.get(f"{PRESTASHOP_URL}/api/{path}", params=params, auth=AUTH, timeout=30)
    r.raise_for_status()
    return r.json()


def api_put(path, resource_key, body):
    r = requests.put(
        f"{PRESTASHOP_URL}/api/{path}",
        params={"output_format": "JSON"}, auth=AUTH,
        json={resource_key: body}, timeout=30,
    )
    r.raise_for_status()
    return r.json()


def all_category_ids(page_size=100):
    ids, offset = set(), 0
    while True:
        data = api_get("categories", params={"display": "full", "limit": f"{offset},{page_size}"})
        rows = data.get("categories") or []
        if not rows:
            return ids
        ids.update(int(row["id"]) for row in rows)
        offset += page_size


def all_products(page_size=100):
    offset = 0
    while True:
        data = api_get("products", params={"display": "full", "limit": f"{offset},{page_size}"})
        rows = data.get("products") or []
        if not rows:
            return
        for row in rows:
            yield row
        offset += page_size


def associated_category_ids(product):
    categories = ((product.get("associations") or {}).get("categories") or {}).get("category") or []
    return [int(row["id"]) for row in categories]


def repair_product_default_category(product_id, new_default_id):
    data = api_get(f"products/{product_id}")
    product = data["product"]
    product["id_category_default"] = new_default_id

    categories = product.setdefault("associations", {}).setdefault("categories", {})
    rows = categories.setdefault("category", [])
    if not any(int(row["id"]) == new_default_id for row in rows):
        rows.append({"id": new_default_id})

    return api_put(f"products/{product_id}", "product", product)


def run():
    valid_category_ids = all_category_ids()
    reassigned = 0
    flagged = 0
    for product in all_products():
        product_id = int(product["id"])
        current_default_id = int(product.get("id_category_default") or 0)
        decision = choose_valid_default_category(
            product_id, current_default_id, associated_category_ids(product),
            valid_category_ids, FALLBACK_ROOT_CATEGORY_ID,
        )
        if decision["action"] == "none":
            continue
        if decision["action"] == "flag_manual":
            flagged += 1
            log.warning("Product id=%s has no valid category to fall back to. Needs manual review.", product_id)
            continue

        log.info(
            "Product id=%s old id_category_default=%s new id_category_default=%s. %s",
            product_id, decision["old_default"], decision["new_default"],
            "would reassign" if DRY_RUN else "reassigning",
        )
        if not DRY_RUN:
            repair_product_default_category(product_id, decision["new_default"])
        reassigned += 1
    log.info("Done. %d product(s) %s, %d flagged for manual review.",
              reassigned, "to reassign" if DRY_RUN else "reassigned", flagged)


if __name__ == "__main__":
    run()
